### PR TITLE
feat(goal_planner): sort goal candidates by weighted distance

### PR DESCRIPTION
**Pull request by kosuke55**
## Description

sort goal candidates priority by weighted distance

image

image

## Tests performed

psim

2023/09/26 https://evaluation.tier4.jp/evaluation/reports/a2e2c74e-2ea8-5f80-80ec-80b1e801e584/?project_id=prd_jt

## Interface changes

none

## Effects on system behavior

none

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/src/utils/goal_planner/goal_searcher.cpp [hunk 1]
```diff
@@ -38,6 +38,33 @@ using lanelet::autoware::NoStoppingArea;
...
+    const double diff = a.distance_from_original_goal - b.distance_from_original_goal;
+    constexpr double eps = 0.01;
+    if (std::abs(diff) < eps) {
+      return a.lateral_offset < b.lateral_offset;
+    }
+    return a.distance_from_original_goal < b.distance_from_original_goal;
+  }
+};
+
+// Sort with weighted lateral distance against longitudinal distance
```
Comment (kyoichi-sugahara) [inline planning/behavior_path_planner/src/utils/goal_planner/goal_searcher.cpp:55]: [NITS]
If my understanding is correct, comments below is easier to understand for me.
Sort with the weighted sum of the longitudinal distance and the lateral distance weighted by lateral_cost.

File: planning/behavior_path_planner/src/utils/goal_planner/goal_searcher.cpp [hunk 2]
```diff
@@ -38,6 +38,33 @@ using lanelet::autoware::NoStoppingArea;
...
 using tier4_autoware_utils::inverseTransformPose;
 
+// Sort with smaller longitudinal distances taking precedence over smaller lateral distances.
+struct SortByLongitudinalDistance
+{
+  bool operator()(const GoalCandidate & a, const GoalCandidate & b) const noexcept
+  {
+    const double diff = a.distance_from_original_goal - b.distance_from_original_goal;
+    constexpr double eps = 0.01;
+    if (std::abs(diff) < eps) {
```
Comment (kyoichi-sugahara) [inline planning/behavior_path_planner/src/utils/goal_planner/goal_searcher.cpp:48]:
```suggestion
　// If the longitudinal distances are approximately equal, sort based on lateral offset.
    if (std::abs(diff) < eps) {
```

**Review by kyoichi-sugahara — APPROVED** [review]
LGTM!

File: planning/behavior_path_planner/src/utils/goal_planner/goal_searcher.cpp (hunk 1 above)
Comment (kosuke55) [reply planning/behavior_path_planner/src/utils/goal_planner/goal_searcher.cpp]: thanks! it is better! fixed in https://github.com/autowarefoundation/autoware.universe/pull/5110/commits/e6f3068375f3246ebaa10c448e03ceb3f2dec7da

**Pull request merged**